burstphoto/align.metal

```diff
@@ -416,3 +740,992 @@ kernel void compute_merge_weight(texture2d<float, access::read> texture_diff [[t
+          
+            // use modified raised cosine window to apply lower weights at outer regions of the patch
+            norm_cosine = (0.5f-0.17f*cos(-angle*((dx+x_shift)+0.5f)))*(0.5f-0.17f*cos(-angle*((dy+y_shift)+0.5f)));
+            
+            tile_diff += norm_cosine * abs(ref_texture.read(uint2(dx, dy)) - aligned_texture.read(uint2(dx, dy)));
+             
+            n_total += norm_cosine;
+        }
+    }
+     
+    tile_diff /= n_total;
+
+    // calculation of mismatch ratio, which is different from the Wiener shrinkage proposed in the publication above (equation (8)). The quadratic terms of the Wiener shrinkage led to a strong separation of bright and dark pixels in the mismatch texture while mismatch should be (almost) independent of pixel brightness
+    float4 const mismatch4 = tile_diff / sqrt(noise_est+1.0f);
+    float const mismatch = 0.25f*(mismatch4[0] + mismatch4[1] + mismatch4[2] + mismatch4[3]);
+        
+    mismatch_texture.write(mismatch, gid);
+}
+
+
+kernel void normalize_mismatch(texture2d<float, access::read_write> mismatch_texture [[texture(0)]],
+                               constant float* mean_mismatch_buffer [[buffer(0)]],
+                               uint2 gid [[thread_position_in_grid]]) {
+    
+    // load args
+    float const mean_mismatch = mean_mismatch_buffer[0];
+    
+    float mismatch_norm = mismatch_texture.read(gid).r;
+    
+    // normalize that mean value of mismatch texture is set to 0.12, which is close to the threshold value of 0.17. For values larger than the threshold, the strength of temporal denoising is not increased anymore
+    mismatch_norm *= (0.12f/(mean_mismatch + 1e-12f));
+    
+    // clamp to range of 0 to 1 to remove very large values
+    mismatch_norm = clamp(mismatch_norm, 0.0f, 1.0f);
+    
+    mismatch_texture.write(mismatch_norm, gid);
+}
+
+
+// hot pixel correction based on the idea that hot pixels appear at the same pixel location in all images
+kernel void correct_hotpixels(texture2d<float, access::read> average_texture [[texture(0)]],
+                              texture2d<float, access::read> in_texture [[texture(1)]],
+                              texture2d<float, access::write> out_texture [[texture(2)]],
+                              constant float* mean_texture_buffer [[buffer(0)]],
+                              uint2 gid [[thread_position_in_grid]]) {
+       
+    int const x = gid.x+2;
+    int const y = gid.y+2;
+    
+    // load args
+    float mean_texture = 0.0f;
+    
+    // extract color channel-dependent mean value of the average texture of all images
+    if (x%2 == 0 & y%2 == 0) {
+        mean_texture = mean_texture_buffer[0];
+    }
+    if (x%2 == 1 & y%2 == 0) {
+        mean_texture = mean_texture_buffer[1];
+    }
+    if (x%2 == 0 & y%2 == 1) {
+        mean_texture = mean_texture_buffer[2];
+    }
+    if (x%2 == 1 & y%2 == 1) {
+        mean_texture = mean_texture_buffer[3];
+    }
```

Comment:
Much more elegant. I will include this change in a future update together with other changes. It will be part of the "exposure branch" as the hot pixel correction was already updated.

Reply:
Due to some changes in the code of the hot pixel correction, this suggestion cannot be implemented as proposed.